### PR TITLE
from_tiff convenience function

### PR DESCRIPTION
**Pull request by tlambert03**
adds a `from_tiff` convenience to generate an `OME` class from a tiff, using `tifffile`.  
Stopped short of adding `tifffile` to dependencies though (added an extra).  Ideas for extracting XML without `tifffile` welcome.

### CODE REVIEW

**Comment by jmuhlich** [comment]
I got you fam: https://github.com/labsyspharm/ome-tiff-pyramid-upgrade/blob/master/pyramid_upgrade.py#L36
I wrote that to read, manipulate, and write IFDs in order to upgrade the old legacy OME-TIFF "Faas" pyramids to the new BioFormats 6 format in place, without reading or writing any of the pixel data so it runs in an instant (and doesn't require space for a copy). With a little more work it could be a general purpose IFD reading library (like adding classic TIFF support and the ability to refer to tags by name instead of number) but I could easily extract just enough code to read only the ImageDescription out of the first IFD. It only uses pure Python.

**Comment by tlambert03** [comment]
> I got you fam

lol... I had to google that :joy:

wow 😱  that looks amazing.  I _think_ I'm all for it if you are!  Though I do want a bit more thorough background on exactly what the difference are.  what the limitations of tifffile are, what motivated you to do this, what potential limitations might be here... 

but yeah. this looks crazy

**Comment by jmuhlich** [comment]
For now, if we just want to extract the ImageDescription it's probably 10-20 lines of code. I can omit all the defensive code (which I needed since I didn't want to corrupt files on write), writing code, and general purpose Tag/TagSet data structures. I'll see what I can scrape together. PR against this branch?

The rest of the code in TiffSurgeon would only be required for a full OME-TIFF read/write library, which is out of scope for ome_types. Or is it...?

**Comment by tlambert03** [comment]
>  PR against this branch?

yeah, I'd like to see what that looks like ... (also feel free to open a new PR if you want)

> The rest of the code in `TiffSurgeon` would only be required for a full OME-TIFF read/write library, which is out of scope for ome_types. Or is it...?

To me, it really doesn't matter if it's _in_ `ome-types` or some other package that plays nicely with ome-types... but it would be great to have a pure python library with a nice API to write a numpy array to ome.tiff.  Again, I wouldn't want to reinvent any wheels (ala tifffile), but if there are new (bioformats 6?) features that don't have a pure-python library yet, and if you've already got a start on that sort of thing, it would be great to hook it up here (whether directly or indirectly)

**Comment by cgohlke** [comment]
>  Ideas for extracting XML without tifffile welcome.

Here's a small standalone function to read the first ImageDescription tag value (as NULL terminated bytes) from an open TIFF file handle: 

```Python
def tiff_image_description(fh):
    """Return value of first ImageDescription tag from open TIFF file."""
    from struct import unpack
    offsetsize, offsetformat, tagnosize, tagnoformat, tagsize, codeformat = {
        b'II*\0': (4, '<I', 2, '<H', 12, '<H'),
        b'MM\0*': (4, '>I', 2, '>H', 12, '>H'),
        b'II+\0': (8, '<Q', 8, '<Q', 20, '<H'),
        b'MM\0+': (8, '>Q', 8, '>Q', 20, '>H'),
    }[fh.read(4)]
    fh.read(4 if offsetsize == 8 else 0)
    fh.seek(unpack(offsetformat, fh.read(offsetsize))[0])
    for _ in range(unpack(tagnoformat, fh.read(tagnosize))[0]):
        tagstruct = fh.read(tagsize)
        if unpack(codeformat, tagstruct[:2])[0] == 270:
            size = unpack(offsetformat, tagstruct[4:4+offsetsize])[0]
            if size <= offsetsize:
                return tagstruct[4+offsetsize:4+offsetsize+size]
            fh.seek(unpack(offsetformat, tagstruct[-offsetsize:])[0])
            return fh.read(size)
```

See also https://docs.openmicroscopy.org/ome-model/6.1.2/ome-tiff/code.html#extracting-a-tiff-comment

**Comment by JacksonMaxfield** [comment]
Resolves #50

**Comment by jmuhlich** [comment]
Thanks @cgohlke! That's about what I was envisioning.  :)

**Comment by tlambert03** [comment]
Thanks @cgohlke!

File: src/ome_types/__init__.py [hunk 1]
```diff
@@ -16,14 +17,71 @@
...
+    """Generate OME metadata object from XML string or path.
+
+    Parameters
+    ----------
+    xml : Union[Path, str]
+        Path to an XML file, or literal XML string.
+
+    Returns
+    -------
+    ome: ome_types.OME
```
Comment (jmuhlich) [inline src/ome_types/__init__.py:33]: Oops, this crept in from your other branch I think. :stuck_out_tongue:

Comment (jmuhlich) [reply src/ome_types/__init__.py:33]: I only noticed because there is a merge conflict. on those 2 lines...

Comment (jmuhlich) [reply src/ome_types/__init__.py:33]: Erm, I see now that `OME` _is_ imported there, we just need to decide which is the proper way to write it there. On introspection it's reported as where it's defined, which is `ome_types.model.ome.OME`.

Comment (tlambert03) [reply src/ome_types/__init__.py]: went with `ome_types.model.ome.OME`

**Pull request merged**